src/frontend/apps/impress/src/features/language/LanguagePicker.tsx

```diff
@@ -33,44 +38,80 @@ const SelectStyled = styled(Select)<{ $isSmall?: boolean }>`
 
 export const LanguagePicker = () => {
   const { t, i18n } = useTranslation();
-  const { preload: languages } = i18n.options;
+  const { toast } = useToastProvider();
+  const { mutateAsync: changeUserLanguage } = useChangeUserLanguage();
+  const { userData } = useAuthStore();
+  const { data: conf } = useConfig();
+
+  // Early return if LANGUAGES is not available or empty
+  if (!conf?.LANGUAGES || conf.LANGUAGES.length === 0) {
+    return null;
+  }
+
+  // Create options for the select component
+  const optionsPicker = conf.LANGUAGES.map(([locale, label]) => ({
+    value: locale,
+    label: label,
+    render: () => (
+      <Box
+        className="c_select__render"
+        $direction="row"
+        $gap="0.7rem"
+        $align="center"
+      >
+        <Text $isMaterialIcon $size="1rem" $theme="primary" $variation="600">
+          translate
+        </Text>
+        <Text $theme="primary" $variation="600">
+          {label}
+        </Text>
+      </Box>
+    ),
+  }));
+
+  // Soft match locale
+  const getMatchingLocale = (): string => {
+    const availableLocales = conf.LANGUAGES.map(([locale]) => locale);
+    return (
+      availableLocales.find(
+        (availableLocale) =>
+          availableLocale === i18n.language ||
+          availableLocale.startsWith(i18n.language.split('-')[0]),
+      ) || availableLocales[0]
+    );
+  };
+
+  // Switch i18n.language and user.language via API
+  const switchLanguage = (targetLocale: string): void => {
+    const actions: Promise<unknown>[] = [i18n.changeLanguage(targetLocale)];
+
+    if (userData?.id) {
+      actions.push(
+        changeUserLanguage({
+          userId: userData.id,
+          language: targetLocale,
+        }),
+      );
+    }
```

Comment:
> I was expecting the language to be set on the user on first logging here: https://github.com/numerique-gouv/impress/blob/main/src/backend/core/authentication/backends.py#L91
> 
> Otherwise how and when is this field set on the user? 🤔

It happens here.
When the user is unsatisfied with the language that is set, he switches it and in turn it updates the attribute at the user.

Reply:
@sampaccoud

Reply:
I thought users are created through the Identity Provider? Imo the IdP should transmit the preferred language for initial Account creation. 

It is not garantueed that the user has visited the frontend before he logs in/creates an account, is it?

Reply:
Ok. It could be one possibility but it's not in the case of our idP. If it came from the idP it would have to be added here right as well right?
If it's not set by the idP, you sould set it from the request at the moment of user creation ie on first loggging like I propose?

Reply:
Ok, i will try to implement it on monday

Reply:
Unfortunately, i did not find an elegant way, with the time i had.
I'm in the holidays now and will revist this in january 2025.

Reply:
You are in [this authenticate method](https://github.com/mozilla/mozilla-django-oidc/blob/main/mozilla_django_oidc/auth.py#L284) called with `request` as parameter so you can get the preferred language as you were doing elsewhere?